File: app/src/main/java/org/cuberite/android/helpers/InstallHelper.java
package org.cuberite.android.helpers;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.net.Uri;
import android.os.Handler;

import org.cuberite.android.helpers.StateHelper.State;
import org.cuberite.android.receivers.ProgressReceiver;
import org.cuberite.android.services.InstallService;

public class InstallHelper {
    public static String getDownloadHost() {
        return "https://download.cuberite.org/androidbinaries/";
    }

    public static void installCuberiteDownload(final Activity activity, State state) {
        SharedPreferences preferences = activity.getSharedPreferences(activity.getPackageName(), Context.MODE_PRIVATE);

        Intent intent = new Intent(activity, InstallService.class)
                .setAction("download")
                .putExtra("downloadHost", getDownloadHost())
                .putExtra("state", state)
                .putExtra("targetFolder", preferences.getString("cuberiteLocation", ""))
                .putExtra("receiver", new ProgressReceiver(activity, new Handler()));

        activity.startService(intent);
    }

    public static void installCuberiteLocal(Activity activity, State state, Uri selectedFileUri) {
        SharedPreferences preferences = activity.getSharedPreferences(activity.getPackageName(), Context.MODE_PRIVATE);

        if (selectedFileUri != null) {
            Intent intent = new Intent(activity, InstallService.class)
                    .setAction("unzip")
                    .putExtra("uri", selectedFileUri)
                    .putExtra("state", state)
                    .putExtra("targetFolder", preferences.getString("cuberiteLocation", ""))
                    .putExtra("receiver", new ProgressReceiver(activity, new Handler()));

            activity.startService(intent);
        }
    }
}

File: app/src/main/kotlin/org/cuberite/android/helpers/InstallHelper.kt
package org.cuberite.android.helpers

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.Handler
import android.os.Looper
import org.cuberite.android.receivers.ProgressReceiver
import org.cuberite.android.services.InstallService

object InstallHelper {
    const val DOWNLOAD_HOST = "https://download.cuberite.org/androidbinaries/"

    fun installCuberiteDownload(activity: Activity, state: StateHelper.State?) {
        val preferences = activity.getSharedPreferences(activity.packageName, Context.MODE_PRIVATE)
        val intent = Intent(activity, InstallService::class.java)
                .setAction("download")
                .putExtra("downloadHost", DOWNLOAD_HOST)
                .putExtra("state", state)
                .putExtra("targetFolder", preferences.getString("cuberiteLocation", ""))
                .putExtra("receiver", ProgressReceiver(activity, Handler(Looper.getMainLooper())))
        activity.startService(intent)
    }

    fun installCuberiteLocal(activity: Activity, state: StateHelper.State?, selectedFileUri: Uri?) {
        val preferences = activity.getSharedPreferences(activity.packageName, Context.MODE_PRIVATE)
        selectedFileUri?.let {
            val intent = Intent(activity, InstallService::class.java)
                    .setAction("unzip")
                    .putExtra("uri", selectedFileUri)
                    .putExtra("state", state)
                    .putExtra("targetFolder", preferences.getString("cuberiteLocation", ""))
                    .putExtra("receiver", ProgressReceiver(activity, Handler(Looper.getMainLooper())))
            activity.startService(intent)
        }
    }
}
